lib/postprocessing/image.js

```diff
@@ -94,113 +93,50 @@ function processBasicImageOperations(img, instructions) {
         img.units('pixelsperinch');
         if (typeof instructions.convertToDpi === 'object') {
             const { xdpi, ydpi } = instructions.convertToDpi;
-            img = img.resample(xdpi, ydpi);
+            img.resample(xdpi, ydpi);
         } else {
-            img = img.resample(instructions.convertToDpi, instructions.convertToDpi);
+            img.resample(instructions.convertToDpi, instructions.convertToDpi);
         }
     }
-
-    console.log('Processing rendition... Done!');
-    return img;
 }
 
-// Watermarking
-function processWatermark(img, watermarkInstructions) {
-    console.log('Processing watermark...');
-    const placement = watermarkInstructions.placement || 'Center';
-    const visibilityPercentage = watermarkInstructions.visibilityPercentage || 100;
-
-    let watermark = watermarkInstructions.watermarkContent;
-    if(watermarkInstructions.isEmbedded){
-        // watermark embedded as base 64
-        console.log('Using embedded watermark');
-        watermark = 'inline:'+watermarkInstructions.watermarkContent;
-    }
-
-    img.dissolve(visibilityPercentage).gravity(placement);
-    img.composite(watermark);
+async function checkJpegSize(outfile, instructions) {
+    // If we specified jpegSize for JPEG output we need to verify that the size does not exceed that
+    if (instructions.fmt === "jpg" && instructions.jpegSize) {
 
-    console.log('Processing watermark... Done!');
-    return img;
-}
+        const stat = await fs.stat(outfile);
+        if (stat.size > instructions.jpegSize) {
+            const img = gm(outfile);
 
-function applyImageOperations(img, instructions, operations){
-    if(operations.basicProcessing) {
-        img = processBasicImageOperations(img, instructions);
-    }
+            // From http://www.imagemagick.org/Usage/formats/#jpg
+            // As of IM v6.5.8-2 you can specify a maximum output filesize for the JPEG image.
+            img.define(`jpeg:extent=${instructions.jpegSize}`);
 
-    if(operations.watermarking) {
-        if(instructions.watermark && typeof instructions.watermark === 'object'){
-            img = processWatermark(img, instructions.watermark);
+            await img.write(outfile);
         }
     }
+}
 
-    if(process.env.ASSET_COMPUTE_NO_METADATA_IN_IMG || operations.removeMetadata){
-        // removes profile and comments, for reproduceable images (for binary file comparision)
-        console.log('Removing profile and metadata from rendition');
-        img = img.strip();
-    }
+async function write(img, outfile, instructions) {
```

Comment:
Return `img` and it will become a fluent API...
(https://en.wikipedia.org/wiki/Fluent_interface)

I find it quite nice to have the postprocessing fluent. Feels more natural for image treatment.

Reply:
See above.